### PR TITLE
add provide precheck and fail fast

### PR DESCRIPTION
**Pull request by zirain**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

- fail fast when secret is wrong
- requeue when secret is updated

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/apis/cluster/v1alpha1/cluster_consts.go [hunk 1]
```diff
@@ -37,10 +37,16 @@ const (
 
 	// ReadyCondition defines the Ready condition type that summarizes the operational state of a Cluster.
 	ReadyCondition capiv1.ConditionType = "Ready"
+	// ProvisioningReason (Severity=Info) documents that the cluster is being provisioned.
+	ProvisioningReason = "Provisioning"
+	// DeletingReason (Severity=Info) documents that the cluster is being provisioned.
+	DeletingReason = "Deleting"
```
Comment (hzxuzhonghu) [inline pkg/apis/cluster/v1alpha1/cluster_consts.go:43]: comment is wrong

File: pkg/apis/cluster/v1alpha1/cluster_consts.go [hunk 2]
```diff
@@ -37,10 +37,16 @@ const (
...
 	// ProviderInitializeFailedReason (Severity=Error) documents that the cluster infrastructure provider initilize failed, probabily a wrong credential.
 	ProviderInitializeFailedReason = "ProviderInitializeFailed"
 	// ClusterResourceSetProvisionFailedReason (Severity=Error) documents that the additinal Cluster API resources (ClusterResourceSet etc.) provisioning failed.
 	ClusterResourceSetProvisionFailedReason = "ClusterResourceSetProvisionFailed"
+	// PrecheckFailedReason (Severity=Error) documents that the cluster precheck is failed.
+	PrecheckFailedReason = "PrecheckFailed"
```
Comment (hzxuzhonghu) [inline pkg/apis/cluster/v1alpha1/cluster_consts.go:49]: can we validate some format error in webhook?

File: pkg/controllers/infra/cluster_controller.go [hunk 3]
```diff
@@ -45,6 +49,10 @@ import (
 	"kurator.dev/kurator/pkg/infra/util"
 )
 
+var (
+	log = ctrl.Log.WithName("cluster-controller")
```
Comment (hzxuzhonghu) [inline pkg/controllers/infra/cluster_controller.go:53]: Do we still need L69?

File: pkg/apis/cluster/v1alpha1/cluster_consts.go (hunk 2 above)
Comment (zirain) [reply pkg/apis/cluster/v1alpha1/cluster_consts.go:49]: yes, we can validate miss/wrong key in webhook.

File: pkg/controllers/infra/cluster_controller.go (hunk 3 above)
Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go:53]: L421 use `log` with name.

File: pkg/apis/cluster/v1alpha1/cluster_consts.go (hunk 2 above)
Comment (zirain) [reply pkg/apis/cluster/v1alpha1/cluster_consts.go:49]: but we do not get a `Secret` with client in webhook.

File: pkg/controllers/infra/cluster_controller.go [hunk 4]
```diff
@@ -369,7 +389,50 @@ func (r *ClusterController) reconcileCNI(scopeCluster *scope.Cluster) error {
...
+		return nil
+	}
+
+	sets := sets.NewString()
+	for _, cluster := range clusters.Items {
+		if cluster.Spec.Credential.SecretRef == s.GetName() {
+			nn := client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Name}
+			if !sets.Has(nn.String()) {
+				result = append(result, ctrl.Request{NamespacedName: nn})
+				sets.Insert(nn.String())
```
Comment (hzxuzhonghu) [inline pkg/controllers/infra/cluster_controller.go:431]: I am concerned about the performance, can we do as what we have done in other places, add a ownerref or label to secret. Then we do not need to loop over all clusters here

Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go:431]: I address my current thinking on L416, at least `OwnerReference` is not a good idea.

File: pkg/infra/aws/provider.go [hunk 5]
```diff
@@ -123,6 +124,21 @@ func awsConfig(region string, cred *AWSCredentials) *aws.Config {
...
+func (p *AWSProvider) Precheck(ctx context.Context) error {
+	sess, err := session.NewSession(p.config)
+	if err != nil {
+		return errors.New("failed to create AWS session")
+	}
+
+	stsSvc := sts.New(sess)
+	_, err = stsSvc.GetCallerIdentity(&sts.GetCallerIdentityInput{})
+	if err != nil {
+		return errors.New("failed to get caller identity")
```
Comment (hzxuzhonghu) [inline pkg/infra/aws/provider.go:136]: add err info

File: pkg/controllers/infra/cluster_controller.go (hunk 3 above)
Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go:53]: in `SetupWithManager`, there's no context

File: pkg/controllers/infra/cluster_controller.go [hunk 6]
```diff
@@ -164,6 +197,44 @@ func (r *ClusterController) deleteClusterResourceSets(ctx context.Context, scope
 	return nil
 }
 
+// reconcileSecretOwnerReference will add owner reference in the begining of
+func (r *ClusterController) reconcileSecretOwnerReference(ctx context.Context, cluster *clusterv1alpha1.Cluster, removeOwnerRef bool) error {
```
[the remark this answers is not in the document]
Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go]: done

File: pkg/controllers/infra/cluster_controller.go [hunk 7]
```diff
@@ -89,16 +100,32 @@ func (r *ClusterController) Reconcile(ctx context.Context, req ctrl.Request) (_
 		}
 	}()
 
+	if err := r.reconcileSecretOwnerReference(ctx, cluster); err != nil {
+		cluster.Status.Phase = string(clusterv1alpha1.ClusterPhaseFailed)
+		l.Error(err, "reconvile secret failed, please check the cluster's configuration")
```
Comment (hzxuzhonghu) [inline pkg/controllers/infra/cluster_controller.go:105]: reconvile typo

File: pkg/controllers/infra/cluster_controller.go [hunk 8]
```diff
@@ -89,16 +100,32 @@ func (r *ClusterController) Reconcile(ctx context.Context, req ctrl.Request) (_
...
 	provider, err := r.newProvider(scope)
 	if err != nil {
 		conditions.MarkFalse(cluster, clusterv1alpha1.ReadyCondition, clusterv1alpha1.ProviderInitializeFailedReason, capiv1.ConditionSeverityError, err.Error())
 		return ctrl.Result{RequeueAfter: r.RequeueAfter}, errors.Wrapf(err, "failed to create Cluster Provider %s/%s", cluster.Namespace, cluster.Name)
 	}
 
+	if err := provider.Precheck(ctx); err != nil {
+		conditions.MarkFalse(cluster, clusterv1alpha1.ReadyCondition, clusterv1alpha1.PrecheckFailedReason, capiv1.ConditionSeverityError, err.Error())
+		cluster.Status.Phase = string(clusterv1alpha1.ClusterPhaseFailed)
+		l.Error(err, "precheck failed, please check the cluster's credential")
```
Comment (hzxuzhonghu) [inline pkg/controllers/infra/cluster_controller.go:120]: we should not mention cluster's credential here. It should be included in err msg

Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go:120]: should not assume credential file exist, it's optional.

File: pkg/controllers/infra/cluster_controller.go (hunk 7 above)
Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go]: fixed

File: pkg/controllers/infra/cluster_controller.go (hunk 8 above)
Comment (hzxuzhonghu) [reply pkg/controllers/infra/cluster_controller.go:120]: I mean you should mention credential in implementor. In the caller you should not assume what's wrong

Comment (zirain) [reply pkg/controllers/infra/cluster_controller.go:120]: done

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**